processing/src/main/java/io/druid/query/QueryMetrics.java

```diff
@@ -0,0 +1,198 @@
+ * Broken builds of custom extensions, containing custom QueryMetrics is the way to notify users that Druid core "wants"
+ * to emit new dimension or metric, and the user handles them manually: if the new dimension or metric is useful and not
+ * very expensive to process and store then emit, skip (see above Goals, 1.) otherwise.
+ *
+ * <p>QueryMetrics is designed for use from a single thread, implementations shouldn't care about thread-safety.
+ *
+ *
+ * Adding new methods to QueryMetrics
+ * ----------------------------------
+ * 1. When adding a new method for setting a dimension, which could be pulled from the query object, always make them
+ * accept a single `QueryType query` parameter, letting the implementations to do all the work of carving the dimension
+ * value out of the query object.
+ *
+ * 2. When adding a new method for setting a dimension, which becomes known in the process of the query execution or
+ * after the query is completed, design it so that as little work as possible is done for preparing arguments for this
+ * method, and as much work as possible is done in the implementations of this method, if they decide to actually emit
+ * this dimension.
+ *
+ * 3. When adding a new method for registering metrics, make it to accept the metric value in the smallest reasonable
+ * unit (i. e. nanoseconds for time metrics, bytes for metrics of data size, etc.), allowing the implementations of
+ * this method to round the value up to more coarse-grained units, if they don't need the maximum precision.
+ *
+ * @param <QueryType>
+ */
+public interface QueryMetrics<QueryType extends Query<?>>
+{
+  /**
+   * Pulls all information from the query object into dimensions of future metrics.
+   */
+  void query(QueryType query);
+
+  /**
+   * Sets {@link Query#getDataSource()} of the given query as dimension.
+   */
+  void dataSource(QueryType query);
+
+  /**
+   * Sets {@link Query#getType()} of the given query as dimension.
+   */
+  void queryType(QueryType query);
+
+  /**
+   * Sets {@link Query#getIntervals()} of the given query as dimension.
+   */
+  void interval(QueryType query);
+
+  /**
+   * Sets {@link Query#hasFilters()} of the given query as dimension.
+   */
+  void hasFilters(QueryType query);
+
+  /**
+   * Sets {@link Query#getDuration()} of the given query as dimension.
+   */
+  void duration(QueryType query);
+
+  /**
+   * Sets {@link Query#getId()} of the given query as dimension.
+   */
+  void queryId(QueryType query);
+
+  /**
+   * Sets {@link Query#getContext()} of the given query as dimension.
+   */
+  void context(QueryType query);
+
+  void server(String host);
+
+  void remoteAddress(String remoteAddress);
+
+  void userDimensions(Map<String, String> userDimensions);
+
+  void status(String status);
+
+  void success(boolean success);
+
+  /**
+   * Registers "query time" metric.
+   */
+  QueryMetrics<QueryType> queryTime(long timeNs);
```

Comment:
I feel like these should have some extra word on them so they don't "look" the same as the other metrics. I recognize they can chain, and that's a difference, but they still look similar to me. Maybe `reportQueryTime` instead of `queryTime`?